### PR TITLE
Finish Stringsdict validation logic

### LINKED ISSUE #6: Use a data structure for problems, not just strings
We should have an explicit list of problems, each with an identifier, and ideally a way for people to silence specific warnings.

### PR DESCRIPTION
**Pull request by stevelandeyasana**
Adds a data structure for error messages to fix #6.

Adds a summary printed to stdout which is much more human-readable than the per-line errors. Stdout is for humans, stderr is for Xcode.

Finds these issues in Stringsdict files:
- Contains no variables (warning)
- Uses implicit positions instead of explicit (ignored, since we have hundreds of these and no plans to fix them)
  - If/when we add config files we can re-enable it and disable it in our own config
- Two arguments in the base language have different specifiers (error)
- The base language skips an argument (warning)
- The translation skips an argument (warning)
- The translation argument has a different specifier than the base language (error)
- A string is missing from a translation (warning)
- The translation uses an argument the original does not have (error if greater than all original arguments, warning if it's earlier than the base language's last argument but the base language doesn't use it so we can't verify its type)

The tests could be improved a lot, but having just fixed a bunch of bugs in Asana's stringsdict files using the tool, I'm happy with the amount of coverage the current test suite provides.

### CODE REVIEW

File: Sources/LocheckLogic/ProblemReporter.swift [hunk 1]
```diff
@@ -15,44 +15,94 @@ private struct StderrOutputStream: TextOutputStream {
...
-
+    public struct LocalProblem: Error, Equatable {
         public let path: String
         public let lineNumber: Int
-        public let message: String
-        public let severity: Severity
+        public let problem: Problem
+
+        var messageForXcode: String {
+            "\(path):\(lineNumber): \(problem.severity.rawValue): \(problem.message) (\(problem.kindIdentifier))"
```
Comment (TheDom) [inline Sources/LocheckLogic/ProblemReporter.swift:47]: Does Xcode show the error on line 1 if we print 0 for the line number here?

File: Sources/LocheckLogic/ProblemReporter.swift [hunk 2]
```diff
@@ -15,44 +15,94 @@ private struct StderrOutputStream: TextOutputStream {
...
 
-        guard log else { return }
+        guard log, problem.severity != .ignored else { return }
         // Print to stderr with formatting for Xcode error reporting
-        print("\(path):\(lineNumber): \(severity.rawValue): \(message)", to: &standardError)
+        print(localProblem.messageForXcode, to: &standardError)
+    }
+
+    public func printSummary() {
+        guard problems.contains(where: { $0.problem.severity == .ignored }) else { return }
```
Comment (TheDom) [inline Sources/LocheckLogic/ProblemReporter.swift]: Should this be `!= .ignored`?

File: Sources/LocheckLogic/Problems.swift [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+struct StringHasMissingArguments: Problem, StringsProblem, Equatable {
+    var kindIdentifier: String { "string_has_missing_arguments" }
+    var uniquifyingInformation: String { "\(language)-\(key)" }
+    var severity: Severity { .warning }
+    let key: String
+    let language: String
+    let args: [String]
+
+    var message: String { "Does not include argument(s) at \(args.joined(separator: ", "))" }
```
Comment (TheDom) [inline Sources/LocheckLogic/Problems.swift:76]: [nit] Consider checking `args.count` to avoid `argument(s)`

**Review by TheDom — APPROVED** [review]
LGTM, congrats!! :)

File: Sources/LocheckLogic/ProblemReporter.swift (hunk 1 above)
Comment (stevelandeyasana) [reply Sources/LocheckLogic/ProblemReporter.swift:47]: No, but it opens the file. Xcode will not show per-line errors for plist files, apparently. That's one reason why I added the summary section, so it's easier for a human to browse the script output.

File: Sources/LocheckLogic/Problems.swift (hunk 3 above)
Comment (stevelandeyasana) [reply Sources/LocheckLogic/Problems.swift:76]: I thought about this but decided to keep all the logic non-branching for generating error messages.

**Pull request merged**